app/uk/gov/hmrc/gvmshaulierstatus/actions/AuthorisedAction.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gvmshaulierstatus.actions
+
+import javax.inject.{Inject, Singleton}
+import play.api.mvc.{ActionBuilder, AnyContent, Request, Results}
+import uk.gov.hmrc.internalauth.client._
+
+import scala.concurrent.Future
+
+@Singleton
+class AuthorisedAction @Inject() (val internalAuth: BackendAuthComponents) {
+
+  private val writePermission: Predicate.Permission = Predicate.Permission(
+    resource = Resource(
+      resourceType = ResourceType("gvms-haulier-status"),
+      resourceLocation = ResourceLocation("*")
+    ),
+    action = IAAction("WRITE")
+  )
+
+  def withInternalAuth: ActionBuilder[Request, AnyContent] =
+    internalAuth.authorizedAction(
+      writePermission,
+      onUnauthorizedError = Future.successful(Results.Unauthorized("Unauthorized request")),
+      onForbiddenError = Future.successful(Results.Forbidden("Forbidden request"))
```

Comment:
if we redesign the class like below we can re-use the write perms in the unit tests and use same Future rather than creating them for each request , what do you think ?

```
trait AuthorisedAction {
  def withInternalAuth: ActionBuilder[Request, AnyContent]
}

object AuthorisedAction {
  val writePermission: Predicate.Permission = Predicate.Permission(
    resource = Resource(
      resourceType = ResourceType("gvms-haulier-status"),
      resourceLocation = ResourceLocation("*")
    ),
    action = IAAction("WRITE")
  )

  val onUnauthorizedError = Future.successful(Results.Unauthorized("Unauthorized request"))
  val onForbiddenError = Future.successful(Results.Forbidden("Forbidden request"))
}

class InternalAuthorisedAction @Inject() (val internalAuth: BackendAuthComponents) extends AuthorisedAction {
  override def withInternalAuth: ActionBuilder[Request, AnyContent] =
    internalAuth.authorizedAction(
      writePermission,
      onUnauthorizedError = onUnauthorizedError,
      onForbiddenError = onForbiddenError
    )
}
```

Reply:
I agree this looks like a nice abstraction but I'm explicitly asserting that `expectedPermissionPredicate` is the one being exercised in this case. If I reference it from the object in my unit tests that assertion will no longer be valid.

Reply:
What I mean is that I'm being explicit in the testing of the permission being exercised if that makes sense.

Reply:
that fine, if you want to have a particular check, you can actually **verify** that if the call is happened with that specific perm or not 
we can still have those perms and futures in the object

Reply:
yes in this case there's a hard `eq` assertion on the stubAuth mock. But I'll add the verify now to make it explicit.